contracts/src/intents/swap/action.rs

```diff
@@ -0,0 +1,43 @@
+use near_sdk::{near, AccountId};
+use serde_with::{serde_as, DefaultOnNull};
+
+use super::{Asset, Deadline, IntentId};
+
+#[derive(Debug, Clone)]
+#[near(serializers = [json, borsh])]
+#[serde(rename_all = "snake_case")]
+pub enum SwapIntentAction {
+    Create(CreateSwapIntentAction),
+    Execute(ExecuteSwapIntentAction),
+}
+
+#[derive(Debug, Clone)]
+#[serde_as]
+#[near(serializers = [json, borsh])]
+pub struct CreateSwapIntentAction {
+    /// Unique ID of intent.  
+    /// NOTE: This MUST not exist before.
+    pub id: IntentId,
+    /// Desired asset as an output.
+    pub asset_out: Asset,
+    /// Where to send asset_out. By default: back to initiator.
+    #[serde(default)]
+    #[serde_as(as = "DefaultOnNull")]
```

Comment:
I personally don't like an approach to introduce new "wrapper" types that are needed exclusively to redefine how the underlying struct is (de)serialized. When working with such types (i.e. `near_sdk::U128`) you have to manually unwrap it by doing `amount.0` and wrap `U128(amount)`, as `U128` doesn't implement all the methods from `u128` primitive type. 

It gets ever worse when you have nested collections:  
Imagine that some function expects `&[u128]` as an input, but you only have `Vec<U128>`. Then, to pass a reference to the slice inside a function that operates only with `u128`s you have to iterate through the whole collection and collect it:
```rust
fn f(slice: &[128]) {
    
}

let wrapped: Vec<U128> = vec![U128(1), U128(2), U128(3)];
let v: Vec<u128> = wrapped.into_iter().map(|v| v.0).collect();
f(v)
```

This seems as a highly discouraged approach to me and it gets worse when you combine it with other types.

Reply:
As for your question:
> What is the difference between #[serde(default)] and #[serde_as(as = "DefaultOnNull")]?

`DefaultOnNull` lets you provide `null` value in JSON representation and it will be treated as `Default::default()` for this type. It makes it less error-prone when other developers wants to use it and forget that `"amount": null` is not the same as absent `"amount"` at all

Reply:
We have what we have by using the `near-sdk`. But I don't see a reason to mix up the approach. Let's do it when we write our own SDK for writing smart contracts. And I don't like additional unnecessary dependencies. At least it could affect the size of the contract.

Reply:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=456442e28fc095669582469b1fb52ed7

Reply:
> We have what we have by using the `near-sdk`.
What exactly makes it a good idea to use an approach that promote a wrong design?

> But I don't see a reason to mix up the approach
I intentionally used `U128` in parts of the code where we interact with `ft_on_transfer`, but I don't see a reason for using this IMHO wrong approach further in the code. From my personal experience it gets really worse when you intentionally propagate wrong ideas and your codebase grows.

> At least it could affect the size of the contract.
As long as we use `lot = true`, the compiler only embeds the code that is used by the target executable, so it doesn't really affects the size of a contract. I agree that having two implementations for `U128` and `DisplayFromStr` that does the same makes it a bit bigger, but  as far as I understand we are not going to have thousands of these contracts deployed, so I would rather compromise with a few bytes, rather then using a wrong approach in terms of design here

Reply:
> https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=456442e28fc095669582469b1fb52ed7

Ah, I see, `Option` indeed handles `null` values under the hood. I'll remove `DefaultOnNull` here.
The thing I was talking about:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=5b2ec5e83967364ee095ac577fab74dc